### PR TITLE
wip

### PR DESCRIPTION
**Pull request by mejango**

### CODE REVIEW

**Comment by mejango** [comment]
@xBA5ED we probably also want to account for the tierBalanceOf in the redemption delegate when burning. maybe the abstract class exposes a `processRedemption` similar to how it does `processContribution` (might rename to processPayment)

**Comment by mejango** [comment]
nevermind!

**Review by drgorillamd — APPROVED** [review]

File: contracts/JBTieredLimitedNFTRewardDataSourceStore.sol [hunk 1]
```diff
@@ -0,0 +1,796 @@
...
+/**
+  @title
+  JBTieredLimitedNFTRewardDataSourceStore
+*/
+contract JBTieredLimitedNFTRewardDataSourceStore is IJBTieredLimitedNFTRewardDataSourceStore {
+  //*********************************************************************//
+  // --------------------------- custom errors ------------------------- //
+  //*********************************************************************//
+
+  error INSUFFICIENT_AMOUNT();
```
[the remark this answers is not in the document]
Comment (mejango) [reply contracts/JBTieredLimitedNFTRewardDataSourceStore.sol:16]: down

File: contracts/JBTieredLimitedNFTRewardDataSourceStore.sol [hunk 2]
```diff
@@ -0,0 +1,796 @@
...
+  mapping(address => uint256) public override numberOfTiers;
+
+  /** 
+    @notice
+    The reward tier data. 
+
+    _nft The NFT contract to which the tier data belong.
+    _tierId The incremental ID of the tier, starting with 1.
+  */
+  mapping(address => mapping(uint256 => JBNFTRewardTierData)) public tierData;
```
[the remark this answers is not in the document]
Comment (mejango) [reply contracts/JBTieredLimitedNFTRewardDataSourceStore.sol:46]: got it

File: contracts/JBTieredLimitedNFTRewardDataSource.sol [hunk 3]
```diff
@@ -918,12 +516,24 @@ contract JBTieredLimitedNFTRewardDataSource is
...
     super._afterTokenTransfer(_from, _to, _tokenId);
   }
+
+  /** 
+    @notice
+    The stored base URI. 
+
+    @return The URI.
+  */
+  function _baseURI() internal view virtual override returns (string memory) {
```
[the remark this answers is not in the document]
Comment (mejango) [reply contracts/JBTieredLimitedNFTRewardDataSource.sol]: I think not

File: contracts/JBTieredLimitedNFTRewardDataSourceStore.sol [hunk 4]
```diff
@@ -0,0 +1,804 @@
...
+  /** 
+    @notice
+    Adds tiers. 
+
+    @param _tierData The tiers to add.
+    @param _constructorTiers A flag indicating if tiers with voting units and reserved rate should be allowed.
+
+    @return tierIds The IDs of the tiers added.
+  */
+  function recordAddTierData(JBNFTRewardTierData[] memory _tierData, bool _constructorTiers)
```
[the remark this answers is not in the document]
Comment (mejango) [reply contracts/JBTieredLimitedNFTRewardDataSourceStore.sol:390]: nice

**Pull request merged**